File: 20-easy-gemstones/gemstones.cpp
#include <bits/stdc++.h>

#include <unordered_set>
#include <unordered_map>

using namespace std;

string ltrim(const string &);
string rtrim(const string &);

/*
 * Complete the 'gemstones' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts STRING_ARRAY arr as parameter.
 */

int gemstones(vector<string> arr) {
    typedef unordered_set <char> MySetType;
    typedef unordered_map <char,size_t> MyMapType;

    MySetType mset;
    MyMapType mmap;

    for(const auto &str : arr){
        for(const auto &c : str){
            mset.insert( c );
        }
        for(const auto &si : mset){
            MyMapType::const_iterator got = mmap.find(si);
            if( got == mmap.end() ){
                mmap[si] = 1;
            }else{
                mmap[si] += 1;
            }
            // mmap[si] = (got == mmap.end()) ? 1 : mmap[si] + 1;
        }
        mset.clear();
    }
    int count=0;
    for(const auto &mi : mmap){
        if( mi.second == arr.size()) ++count;
    }
    
    return count;
}

int main()
{

    string n_temp;
    getline(cin, n_temp);

    int n = stoi(ltrim(rtrim(n_temp)));

    vector<string> arr(n);

    for (int i = 0; i < n; i++) {
        string arr_item;
        getline(cin, arr_item);

        arr[i] = arr_item;
    }

    int result = gemstones(arr);

    if(getenv("OUTPUT_PATH")){
        ofstream fout(getenv("OUTPUT_PATH"));
        fout << result << "\n";
        fout.close();
    }else{
        cout << result << "\n";
    }

    return 0;
}

string ltrim(const string &str) {
    string s(str);

    s.erase(
        s.begin(),
        find_if(s.begin(), s.end(), not1(ptr_fun<int, int>(isspace)))
    );

    return s;
}

string rtrim(const string &str) {
    string s(str);

    s.erase(
        find_if(s.rbegin(), s.rend(), not1(ptr_fun<int, int>(isspace))).base(),
        s.end()
    );

    return s;
}

File: 20-easy-gemstones/gemstones.cs
using System.CodeDom.Compiler;
using System.Collections.Generic;
using System.Collections;
using System.ComponentModel;
using System.Diagnostics.CodeAnalysis;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.Serialization;
using System.Text.RegularExpressions;
using System.Text;
using System;

class Result
{

    /*
     * Complete the 'gemstones' function below.
     *
     * The function is expected to return an INTEGER.
     * The function accepts STRING_ARRAY arr as parameter.
     */

    public static int gemstones(List<string> arr)
    {
        HashSet<char>         mset = new HashSet<char>();
        Dictionary<char, int> mmap = new Dictionary<char, int>();

        foreach(string str in arr){
            foreach(char c in str){
                mset.Add( c );
            }
            foreach(char c in mset){
                if( mmap.ContainsKey(c) ){
                    mmap[c] += 1;
                }else{
                    mmap[c] = 1;
                }
                // mmap[c] = mmap.ContainsKey(c) ? mmcap[c] + 1 : 1;
            }
            mset.Clear();
        }
        int count=0;
        foreach(var mi in mmap){
            if( mi.Value == arr.Count() ){
                ++count;
            }
        }
        return count;
    }

}

class Solution
{
    public static void Main(string[] args)
    {
        TextWriter textWriter = new StreamWriter(@System.Environment.GetEnvironmentVariable("OUTPUT_PATH"), true);

        int n = Convert.ToInt32(Console.ReadLine().Trim());

        List<string> arr = new List<string>();

        for (int i = 0; i < n; i++)
        {
            string arrItem = Console.ReadLine();
            arr.Add(arrItem);
        }

        int result = Result.gemstones(arr);

        textWriter.WriteLine(result);

        textWriter.Flush();
        textWriter.Close();
    }
}
